### DISCUSSION TITLE
.27 point release

### DISCUSSION THREAD

**Discussion by 3point14159** [horizontalsystems/unstoppable-wallet-android#5367] [General]
Thanks to the devs for all the work over the past weeks! Lots of issues addressed and some excellent improvements towards greater usability.

Some of these fixes will allow us to use full character passwords with our multi-lang BIP39 seeds. Hopefully support for watching xpub addresses also.

Considering the number of changes, can we get a point release so we can start using the Unstoppable wallet?

Thanks as always, looking forward to the Unstoppable journey. Great work!

**Comment by 3point14159** [comment]
So watching the development, it seems like much of the issues that are holding-up/stopping a working release are NFT's. Who cares that much about glorified jpegs in-lieu of an actual working app? I haven't been able to use your wallet for over a month after reporting a simple bug of no special characters for BIP39 passphrase, which was fixed long ago. How come it seems so difficult to make a point release for things that have been fixed so people can start using the wallet?

**Comment by 3point14159** [comment]
From the homepage:

“Unstoppable is the first multi-coin wallet that supports Bitcoin, Ethereum, and fully shielded Zcash, as well as other coins, and it has a strong, user-centric architecture in which the users own their own keys and their own privacy."

Zooko Wilcox-O'Hearn
CEO of the Electric Coin Company (ECC)

From me:

We can say that either the CEO of ECC doesn't use a strong passphrase over his seed phrase, or he hasn't tested the basic functionality of the product he's endorsing. Not sure which is worse news...

Also, that quote is untrue. If you go to any other non-custodial wallet's webpage, they all explain how they allow you to retain your own keys and privacy while supporting multi-coins. Who thought this was a great quote to feature on the homepage?

**Comment by esengulov** [comment]
The support for the missing special charaters been implemented. Sorry for the delay. 

The NFTs are here to stay we have been adding the features for it incrementally. Sorry for making you wait.

**Reply by 3point14159** [reply]
Are you just stopping little ole me from using your wallet with a passphrase, or could you possibly imagine there are other people out there that might have a BIP39 passphrase they may need this update also?

You also mentioned the intention of making a release by the end of the week, you guys failed there again too; end of the week is here and no release. No one is surprised, it's expected at this point. But hey, at least your bored yachts make number-go-up in wallet that refuses to get updated. Keep working on NFT's, we'll just keep waiting and testing other wallets in the meantime. Have another fun 3-day weekend while people still can't use your NFT hiatus wallet!

**Reply by esengulov** [reply]
:) we have moved non-started NFT related issues to the next sprint. @Dianaaiym try to push the release ready version asap without waiting for iOS review

**Reply by 3point14159** [reply]
iOS? I thought we were talking about the android version (hence posting here, in the android section). Y'all okay over there? :-|

**Reply by esengulov** [reply]
What I meant to say is publish android version without waiting for the iOS. Usually we push both at the same time.

**Reply by 3point14159** [reply]
[The iOS source got posted as a pre-release](https://github.com/horizontalsystems/unstoppable-wallet-ios/releases/tag/0.27). Seems like Android will get a release this week. Thanks for making it happen (finally) dev team.

**Comment by 3point14159** [comment]
Thanks for pushing the rest of the pending NFT work to the next release, [looks like](https://github.com/horizontalsystems/unstoppable-wallet-android/projects/32) we could have a release today or tomorrow. Would love to see [eCash support in upcoming releases](https://github.com/horizontalsystems/unstoppable-wallet-android/issues/5431). Thanks for the code. Love.

**Reply by 3point14159** [reply]
Congratulations on the [.27 version release today](https://github.com/horizontalsystems/unstoppable-wallet-android/releases/tag/0.27.0)! Can we get an apk posted as an asset here on Github?

**Reply by rafaelekol** [reply]
I attached apk to Release.

**Comment by 3point14159** [comment]
There is **so much** to do to make this mostly functional for just for legacy coins, that I am now even more disappointed to see the energy-sucking NFT focus.

No importing or sweeping of private keys. There is so much you can do with HD BIP39 to make wallets awesome, it's almost like devs are **_actively avoiding_** making their wallets functional to protocol.

In other words; imagine you have libbittorrent, but only design clients to do single downloads only, and limiting the swarm to 2 peers. Hypecoins ruin the ecosystem once again, yesterday it was ICO's, now we have NFT's making devs too dizzy to focus on making a better base product. NFT's have found _a way_ of spamming the blockchains that were designed to be spam-resistant.

**Comment by 3point14159** [comment]
By the way, there is a cool little open-source tool for deriving BIP39 info for many coins. You may want to try to integrate it if you'd like to support most coins out there with one single change/overhaul.

https://github.com/iancoleman/bip39

Best wishes getting your functionality up to speed with the rest of the other multi-coin wallets out there.

**Comment by ealymbaev** [comment]
@3point14159 

I have reviewed all your suggestions regarding restore using various private key types. This feature was not in our schedule actually, that is why it was not implemented yet.

I am going to schedule it into next sprint (moreover it will be the only feature in this short sprint) and here is the list of changes we are planning to apply:

- restore via `EVM Private Key` (e.g. `0cf5e15dee8debe3de5a04a1e228af1a536c3f87bc5e028f04c5eb4c257d8703`)

- restore via `BIP32 Root Key`, including all three BIP44, BIP49, BIP84 (e.g. `xprv9s21ZrQH143K2T6JLi6sPXj3rDjh4FZ88aUSEKXLG4W4fvFQVFn1EyxVQFmEJioG43Hf4ziPMe2PetJDGgNdVSDWNkBcEn7sVs7CxMbLP95`)

- restore via `Account Extended Private Key`, including all three BIP44, BIP49, BIP84 (e.g. `xprv9z7EHnP3BurUqj7AqohtrGU7mTcZ5zNuN5jR1izt5PhPzTmsoEDg4Mxv8VwTMCmDegqWdR3nGPUXM7CWwMkvu6SJhX4RbWB6YduK93tnTos`, `yprvAHVsuzKn9L8FU4wRrWQUzDLznzccEiz6z5SeiFbGcGC6rY5rFwmBmNufvb6PC4pPyNF8SPLD3kgjjgCBp8kU4Lei5B42FLY1mrKA97kY8V9`, `zprvAdKDaL5fSwjJ5y5w3rKS15KACm8BifhPFnZtksAwzGHMZsPYRKfiWFntdvoM6ESwGawxFxu8HbGsLodYSTt51cEY2Y6XFsd6EZozgxrak2k`)

- watch via `Account Extended Public Key`, including all three BIP44, BIP49, BIP84

- also users will be able to view (copy) all of the keys listed above in Settings screens of app (where applicable)

- QR scanner will be added to all Private Key inputs as well

The reason why these features had not been implemented initially is that in certain restore type we need to hide (filter out) non-suitable blockchains from UI (Coin Manager, Wallet Connect and other places). That is why this feature was postponed for later releases, but with current app architecture we can add this type of features with less effort.

Please let me know if this pack of restore types would be enough, or should we consider more options.

**Reply by 3point14159** [reply]
@ealymbaev Thanks for the detailed reply. I am excited to see this is on the roadmap! Do you guys have room for a volunteer to help draft future roadmap objectives so website visitors can learn of these essential inclusions that will set this wallet apart from all others. I have be kicking around developing a watch only wallet using xpub key data. Many people need to track assets outside their local wallet. No one if filling this use-case, and I think many would love the features and further mature crypto as a viable exchange product for the rest of the world. Just thinking big. :)

**Reply by ealymbaev** [reply]
@3point14159 

Please let me know if `BIP32 Root Key`, `Account Extended Private Key` and `Account Extended Public Key` (watch-only) support would be enough. In other words we are not going to support `BIP32 Extended Private/Public Key` because it allows any derivation depth and path - and it leads to a lot of changes in `BitcoinCore` because of `external` and `change` address derivation paths.

Thus `BIP32 Root Key` will be used for restoring `Bitcoin`, compatible Bitcoin forks

`Account Extended Private Key` will allow restoring certain BIP for `Bitcoin` and compatible Bitcoin forks

`Account Extended Public Key` will allow watching certain BIP for `Bitcoin` and compatible Bitcoin forks

P.S. As about volunteer suggestion, let me inform @esengulov for him to reply you

**Reply by 3point14159** [reply]
@ealymbaev  That's a great start. If those features can get integrated, the wallet may be a step above the rest. Would love to contribute to help the usability be easy enough for the general public.

**Reply by ealymbaev** [reply]
@3point14159 

According to this SLIP: https://github.com/satoshilabs/slips/blob/master/slip-0132.md, `xprv` and `xpub` keys are only used for `Bitcoin` blockchain. Do you think it would be a good idea to use them for `BCH` and `DASH` as well? Or should we restore and watch `Bitcoin` blockchain only?

**Reply by 3point14159** [reply]
@ealymbaev 

Some blockchain explorers let you search xpubs:

https://litecoinblockexplorer.net/xpub/Ltub2ZMF3dkMhuMyysQRK53zpwQweBLP3j93cDKWUH933ZfRJoFLjM3yqbVM9DffQgHLVufgCmhPiHeJLoZWNyUt7A7oaG6ZkR8LM7d9DVSvw8V

...and some don't:

https://blockchair.com/search?q=Ltub2ZMF3dkMhuMyysQRK53zpwQweBLP3j93cDKWUH933ZfRJoFLjM3yqbVM9DffQgHLVufgCmhPiHeJLoZWNyUt7A7oaG6ZkR8LM7d9DVSvw8V&submitButton=

Maybe allowing users to export xpub keys for all blockchains, and then also allow them to save links to explorers where they can simply plug in the key/address to view the xpub data. I would be willing to volunteer help with getting websites that have an api you can call to display the info in-app. Thanks for the consideration and responsiveness.

**Reply by ealymbaev** [reply]
@3point14159 

We have almost finished requested features on iOS side. Android is being developed and requires a little more time.

Let me provide some teasers for you to review and comment if required:


https://user-images.githubusercontent.com/377809/197168519-833901de-a682-4de8-8c2f-f261f2fecd01.MP4


https://user-images.githubusercontent.com/377809/197168584-bf60400d-793e-4faf-979c-1cb8a58d2c44.MP4


https://user-images.githubusercontent.com/377809/197168609-0c00a030-072d-4188-a35c-ea02ce7e5ce9.MP4

**Reply by 3point14159** [reply]
@ealymbaev Those look great! Imagine you have xpubs from many wallets that you'd like to watch only; this feature is the perfect! This is a unique feature that you can market against the competition. Thanks for the efforts, now let's help get the android team up to par. I am learning Kotlin so I can start being part of the solution instead of a simple user.

**Reply by 3point14159** [reply]
"Special thanks goes to a number of users pushing us to extend the wallet to cover those use cases ;)"

@ealymbaev Thanks for the complements in the .28 release notes regarding those of us pushing for better utility of the app. Looking forward to more development on xpubs for all possible chains (might have some issues with privacy-only coins) and more coin support (eCash XEC please).
